### PR TITLE
Add a blog entry about KUDO dependencies

### PR DESCRIPTION
**Pull request by zen-dog**
Signed-off-by: Aleksey Dukhovniy <alex.dukhovniy@googlemail.com>

### CODE REVIEW

File: content/blog/blog-2020-07-01-kudo-dependencies.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
+---
+date: 2020-07-01
+---
+
+# KUDO Dependencies
+
+One of the most requested features (internally and externally) has been operator dependencies. We've experienced this pain ourselves: our own [Kafka](https://github.com/kudobuilder/operators/tree/master/repository/kafka) operator requires the user to manually [install Zookeeper](https://github.com/kudobuilder/operators/blob/master/repository/kafka/docs/v1.3/install.md) as a prerequisite. This sounds simple enough but can get out of hand quickly as our [Flink demo](https://github.com/kudobuilder/operators/tree/master/repository/flink/docs/demo/financial-fraud) shows.
+
+We've been thinking about the best way of introducing dependencies to KUDO for a while now. A few months ago we created [KEP-29](https://github.com/kudobuilder/kudo/blob/main/keps/0029-operator-dependencies.md) that aimed at solving at least part of the problems. Dependencies can be a complex topic. KEP-29 is not trying to boil the dependency ocean but rather limits itself to installation dependencies only, i.e. your operator instance and all its dependencies (including transitive ones) will be installed and/or removed as one unit.
```
Comment (kensipe) [inline content/blog/blog-2020-07-01-kudo-dependencies.md]: "at least part of the problems" -> I would lean towards either "part of the problem" or "part of the set of problems".

Comment (kensipe) [reply content/blog/blog-2020-07-01-kudo-dependencies.md]: I think I would be more direct with " Dependencies can be a complex topic" -> "Dependencies is a complex topic"

File: content/blog/blog-2020-07-01-kudo-dependencies.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+<!-- more -->
+
+## KudoOperator Task
+
+KUDO operators already have a mechanism to deal with installation dependencies called [plans, phases, and steps](plans.md) with serial or parallel execution strategy. This mechanism is already powerful enough to express any dependency hierarchy including transitive dependencies. As of KUDO 0.15.x, KUDO supports a `KudoOperator` task which allows you to specify a dependency on other operators. At the very least it requires the name of the operator to be installed. You can also provide an operator version (defaults to the latest one).
```
Comment (kensipe) [inline content/blog/blog-2020-07-01-kudo-dependencies.md]: it looks like `[plans, phases, and steps](plans.md)` was intended to be a link... however the link "plans.md" doesn't go anywhere... did you intend to?

File: content/blog/blog-2020-07-01-kudo-dependencies.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+              - deploy-zookeeper
+      - name: main
+        strategy: parallel
+        steps:
+          - name: second
+            tasks:
+              - deploy-main
+```
+
+As with other tasks, KUDO will make sure this task is healthy before moving to the next one. _"Healthy"_, in the case of the operator means that its `deploy` plan has finished successfully. Note, that any transitive dependencies of the zookeeper operator itself will also be resolved and installed in the right order. `KudoOperator` closely mimics the `kubectl kudo install` command semantics allowing to additionally specify the `appVersion` (defaults to most recent one) and the `instanceName` (will be generated by KUDO by default).
```
Comment (kensipe) [inline content/blog/blog-2020-07-01-kudo-dependencies.md]: "allowing to additionally specify the " is hard to parse IMO... "allowing for the configuration of "?

File: content/blog/blog-2020-07-01-kudo-dependencies.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+$ kubectl kudo install parent -p CHILD_PASSWORD=secret
+```
+
+Note, that the _parent_ operator may decide to provide a sensible default or even to hardcode the password and not expose it at all to the end user.
+
+## Summary
+
+Obviously, there is more to dependencies than the right installation order. One area that KEP-29 deliberately avoided is life-cycle dependencies. It is easy to imagine a situation when a new operator (e.g Kafka) may want to depend on the existing Zookeeper instance. However, such life-cycle dependencies presents major challenges e.g. what happens when Zookeeper is removed? What happens when Zookeeper is upgraded, and the new version is incompatible with the current Kafka Instance? How can we ensure the compatibility? Anyway, we want to gather feedback before we address these issues so ping us if you have any ideas. For more information and implementation details, take a look at the [KEP-29](https://github.com/kudobuilder/kudo/blob/main/keps/0029-operator-dependencies.md), updated [Flink demo](https://github.com/kudobuilder/operators/pull/279) or [dependencies documentation](https://kudo.dev/docs/developing-operators/tasks.html#kudooperator-task).
```
Comment (kensipe) [inline content/blog/blog-2020-07-01-kudo-dependencies.md]: I would remove "Anyway"... it seems dismissive... the sentence works without it.

**Review by kensipe — APPROVED** [review]
lgtm

no blockers.... just suggestions... nice post!

File: content/blog/blog-2020-07-01-kudo-dependencies.md (hunk 1 above)
Comment (ANeumann82) [reply content/blog/blog-2020-07-01-kudo-dependencies.md]: > "at least part of the problems" -> I would lean towards either "part of the problem" or "part of the set of problems".

"part of the problem" sounds better to my ears.

Comment (ANeumann82) [reply content/blog/blog-2020-07-01-kudo-dependencies.md]: > I think I would be more direct with " Dependencies can be a complex topic" -> "Dependencies is a complex topic"

Maybe "dependency handling is a complex topic"? "Dependencies is a complex topic" sounds weird to my non-native ears, I want to make it "Dependencies are a complex topic"...

File: content/blog/blog-2020-07-01-kudo-dependencies.md (hunk 3 above)
Comment (ANeumann82) [reply content/blog/blog-2020-07-01-kudo-dependencies.md]: I agree - Maybe split this sentence?

KudoOperator closely mimics the kubectl kudo install comand semantics. It allows to specify the appVersion ...

File: content/blog/blog-2020-07-01-kudo-dependencies.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+## Dependency Parametrization
+
+Overall we wanted to encourage operator composition by providing a way of operator encapsulation. In other words, operator users should not be allowed to arbitrarily modify the parameters of embedded operator instances. The higher-level operator should define all parameters that its direct dependency operators need. If a child operator needs to be parametrised a parameter file can be specified using the `parameterFile` field. Let's take a look at an example. Suppose, we have two operators (_parent_ and _child_) and _child_ has a required parameter `PASSWORD` which is empty by default. First, the parent operator needs to specify a  parameter file for the child and reference it in the corresponding `KudoOperator` task:
```
Comment (ANeumann82) [inline content/blog/blog-2020-07-01-kudo-dependencies.md]: wanted -> want

Can we use some other parameter than `PASSWORD`? Maybe `USERNAME`? I'd like to get rid of the idea that passwords should be stored in Params...

**Review by ANeumann82 — APPROVED** [review]
lgtm, some nits

File: content/blog/blog-2020-07-01-kudo-dependencies.md (hunk 2 above)
Comment (zen-dog) [reply content/blog/blog-2020-07-01-kudo-dependencies.md]: Good catch 👍

File: content/blog/blog-2020-07-01-kudo-dependencies.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
+---
+date: 2020-07-01
+---
+
+# KUDO Dependencies
+
+One of the most requested features (internally and externally) has been operator dependencies. We've experienced this pain ourselves: our own [Kafka](https://github.com/kudobuilder/operators/tree/master/repository/kafka) operator requires the user to manually [install Zookeeper](https://github.com/kudobuilder/operators/blob/master/repository/kafka/docs/v1.3/install.md) as a prerequisite. This sounds simple enough but can get out of hand quickly as our [Flink demo](https://github.com/kudobuilder/operators/tree/master/repository/flink/docs/demo/financial-fraud) shows.
```
Comment (nfnt) [inline content/blog/blog-2020-07-01-kudo-dependencies.md]: Nitpick, or maybe not a nitpick in CNCF context: Who is "us" when you mention "our own" operator here? IMO, we have to be careful to distinguish the ownership here: The Kafka operator is an [open-source project by D2iQ](https://github.com/mesosphere/kudo-kafka-operator) published in the community repository of CNCF-owned KUDO. Even though the same team at D2iQ works on both projects, we have to distinguish in the external ownership.

Comment (zen-dog) [reply content/blog/blog-2020-07-01-kudo-dependencies.md]: That's an excellent point! I'll change the wording

**Review by nfnt — APPROVED** [review]
